cmd/filebeat-keeper/main.go

```diff
@@ -1,75 +1,98 @@
 	"flag"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"strconv"
 	"sync"
-	"syscall"
 	"text/template"
+	"time"
 
 	"gopkg.in/yaml.v2"
 
 	"github.com/caicloud/logging-admin/pkg/util/graceful"
 	"github.com/caicloud/logging-admin/pkg/util/osutil"
 
 	"github.com/caicloud/nirvana/log"
-	"gopkg.in/fsnotify/fsnotify.v1"
 )
 
 var (
-	filebeatExecutablePath = osutil.Getenv("FB_EXE_PATH", "filebeat")
-	srcConfigPath          = osutil.Getenv("SRC_CONFIG_PATH", "/config/filebeat-output.yml")
-	dstConfigPath          = osutil.Getenv("DST_CONFIG_PATH", "/etc/filebeat/filebeat.yml")
+	filebeatExecutablePath     = osutil.Getenv("FB_EXE_PATH", "filebeat")
+	srcConfigPath              = osutil.Getenv("SRC_CONFIG_PATH", "/config/filebeat-output.yml")
+	dstConfigPath              = osutil.Getenv("DST_CONFIG_PATH", "/etc/filebeat/filebeat.yml")
+	livenessProbePeriodSeconds = osutil.Getenv("LIVENESS_PROBE_PERIOD_SECONDS", "10")
 )
 
-// When configmap being created for the first time, following events received:
-// INFO  1206-09:38:39.496+00 main.go:41 | Event: "/config/..2018_12_06_09_38_39.944532540": CREATE
-// INFO  1206-09:38:39.496+00 main.go:41 | Event: "/config/..2018_12_06_09_38_39.944532540": CHMOD
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/filebeat-output.yml": CREATE
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/..data_tmp": RENAME
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/..data": CREATE
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/..2018_12_06_09_37_32.878326343": REMOVE
-// When configmap being modified, following events received:
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..2018_12_06_09_42_56.160544363": CREATE
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..2018_12_06_09_42_56.160544363": CHMOD
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..data_tmp": RENAME
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..data": CREATE
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..2018_12_06_09_38_39.944532540": REMOVE
-func watchFileChange(path string, reloadCh chan<- struct{}) error {
-	w, err := fsnotify.NewWatcher()
+func hashFile(path string) (string, error) {
+	f, err := os.Open(path)
 	if err != nil {
-		return err
+		return "", err
 	}
-	if err := w.Add(path); err != nil {
-		return err
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", err
 	}
 
-	for {
-		select {
-		case ev := <-w.Events:
-			log.Infoln("Event:", ev.String())
-			if ev.Op&fsnotify.Create == fsnotify.Create {
-				if filepath.Base(ev.Name) == "..data" {
-					log.Infoln("Configmap updated")
-					reloadCh <- struct{}{}
-				}
-			}
-		case err := <-w.Errors:
-			log.Errorf("Watch error: %v", err)
+	return string(h.Sum(nil)), nil
+}
+
+func watchFileChange(path string, reloadCh chan<- struct{}) {
+	var (
+		curHash string
+		mtx     sync.Mutex
+	)
+
+	curHash, err := hashFile(path)
```

Comment:
You created a shadowed `curHash` here due to `:=`, which always create new variables; you could declare the `err` in the variable section, and then do `curHash, err = hashFile(path)` to avoid shadowed variable.

Reply:
fixed